### DISCUSSION TITLE
Paralelisation explanation

### DISCUSSION THREAD

**Discussion by Kubiczek36** [ImSwitch/ImSwitch#209] [Show and tell]
Hi guys, I would need to run couple things in paralell in ImSwitch, however I can't anywhere find explanation, how to do it. would there be anyone to help me? @kasasxav @beniroquai @jonatanalvelid

**Comment by jacopoabramo** [comment]
Hi @Kubiczek36 , ImSwitch provides an internal framework that supports multithreading. Currently ImSwitch takes advantage of Qt for multithreading; @beniroquai worked on a way to detach it from this but it's still a WIP.

There are various examples of how this is achieved in the current codebase; example implementations can be found in the [`FFTController`](https://github.com/ImSwitch/ImSwitch/blob/master/imswitch/imcontrol/controller/controllers/FFTController.py) and [`FocusLockController`](https://github.com/ImSwitch/ImSwitch/blob/master/imswitch/imcontrol/controller/controllers/FocusLockController.py) classes. 

Moreover, could you elaborate a bit further what kind of parallelization do you need?

**Comment by beniroquai** [comment]
@Kubiczek36 would you mind to quickly describe what you want to do exactly? ImSwitch is "just" a python application so you could do anything you would do in another python program too :)

**Reply by Kubiczek36** [reply]
Yeah sure!

Basically here I have an SLM wchich can quickly change the patterns (while triggered)  and we want to make use of it so far for just moving the beam. In order to do that you need 
1. create the phase patterns for the SLM and upload them to the SLM controller
2. choose which pattern you want to be displayed
So I create a 3D stack, upload it there and than I call a method, where I provide an index of phase pattern to be displayed. The SLM than waits for the trigger. (So far in my messy implementation I use the triggering line for the laser).

Just for my SLM I wrote a separate trio of Widget-Controller-Manager, which is a modification of the original SLM implementation. 

Here is the diagram of what happends esentially. Dasshed lines are communications using the signals, solid lines are calls of an function.  
```mermaid
sequenceDiagram
   actor User   
participant recC as RecordingController
participant recM as RecordingManager
participant Cam as CamManager
participant NIDAQ as NIDAQManager + card
participant supScan as Super Scan Controller
participant scanC as Scan Controller 
participant scan as ScanManager   
participant posM as positionersManager 
participant SLMc as SLMCotroller
participant SLM as SLM Manager

note over User : I want to scan 
note over User : pressing REC button

User ->> recC : record while scanning 
recC -->>+ supScan : getNumScan positions
supScan -->>- recC : num scan positions
note over recC : Wait for 0.1 s
recC -->>+ supScan : calls runScanExternal
supScan ->>+ scanC : run scan advanced 
scanC -->> SLMc : createAndUploadScanStack
SLMc ->> SLM : create and upload stack
SLMc -->> scanC : runSLMScan
SLMc  ->> SLM : iterate trough the scan stack
scanC -->> recC : start recording external
recC ->> recM: start recording
note over scanC : puting time.sleep(0.3) here didn't help 
scanC ->> NIDAQ : runScan
```
My problem is, that I would ideally need make sure that I start Iterating trough the SLM stack, just before the NIDAQ runs the triggers, which is not really happening.
```
2024-07-02 11:54:47 DEBUG [NidaqManager] Starting Nidaq scan!
2024-07-02 11:54:47 INFO [NidaqManager] Nidaq scan started!
2024-07-02 11:54:47 DEBUG [SLM_PCIeManager] Iterating through scan stack
2024-07-02 11:54:47 DEBUG [SLM_PCIeManager] Stack length: 11
```
I was thinking that this could be achieved by ensuring that the SLM would run on a separate thread . But I am open to any suggestions from you.

Thanks a lot for your help anyway 🙏🏻

**Reply by jacopoabramo** [reply]
Could you post a link where you implemented your Widget-Controller-Manager trio? I'd like to see the code to get a better understanding of your problem

**Reply by beniroquai** [reply]
You could use asyncio as an alternative. But as far as I understood you have a master (laser or camera) that governs the triggers, right? So you need to prepare the SLM and then synchronize devices using a hardware signal. Why would you need different threads? Maybe (certainly) I'm dumb

**Reply by Kubiczek36** [reply]
@jacopoabramo sure, here you go:
- [Manager](https://github.com/Balzarotti-Lab/ImSwitch/blob/device-implementation/imswitch/imcontrol/model/managers/SLM_PCIeManager.py)
- [Controller](https://github.com/Balzarotti-Lab/ImSwitch/blob/device-implementation/imswitch/imcontrol/controller/controllers/SLM_PCIeController.py)
- [Widget](https://github.com/Balzarotti-Lab/ImSwitch/blob/device-implementation/imswitch/imcontrol/view/widgets/SLM_PCIeWidget.py)

@beniroquai I use NIDAQ to generate the triggers. With the SLM I need to call following line, to pick a frame from a stack (this needs to happen before the NIDAQ starts emitting the signals!). The SLM than waits for the trigger. 
https://github.com/Balzarotti-Lab/ImSwitch/blob/ace7a0d1c9f914b37a2d2465f680aaef2f400c34/imswitch/imcontrol/model/managers/SLM_PCIeManager.py#L256
The camera works a bit differently, it just waits until it gets a trigger and than I save the frame into a queue, from where the REC picks it up. 
Also, even if I am not scanning with the triggers, but rather with just the SLMs timings (it all happens in the SLM widget-controller-manager), than the cam view does not get refreshed. as long as the processes within the SLM trio run. 
I needed a lot PCIe lines, therefore I have beast with Intel Xeon and many cores, so I can afford to use them.

**Reply by Kubiczek36** [reply]
Also, @jacopoabramo we could arrange a call, where you can see how it runs in action, if you want.

**Comment by xcasas** [comment]
Hi @Kubiczek36 ,

If I understand correctly, I don't see how having the SLM on a separate thread would solve the issue. The nidaqManager already works with threads and the camera display is also a thread, so that should be fixed internally in ImSwitch, right? You can surely make the SLMController have a thread inside if you want. But for me the problem is that you need the processes to wait to each other.

How I would do it is by using signals from the communication channel. I see from your sequence diagram that you are using the REC button. As a first approach I would try to have another button instead, in your SLMWidget, that does the calculations first, and then calls the runScan function (if I remember correctly there was another signal for this in the communication channel). Then in the logic behind your new button you can do whatever you want and make sure that everything is triggered asynchronously. So you can first do the SLM stuff, then call the runScan, etc.

Does that make sense? I hope I'm not missing something.

**Reply by xcasas** [reply]
I would check that first and make it work in this way, and then I would consider if it makes sense to integrate this functionality inside the scanning controller in a general way.

**Reply by Kubiczek36** [reply]
Hi Xavi this is kinda already done in this way. 
* First the fuction to prepare the images is called - `create_scan_stack` and `upload_stack` in the [Manager](https://github.com/Balzarotti-Lab/ImSwitch/blob/device-implementation/imswitch/imcontrol/model/managers/SLM_PCIeManager.py)
* once this function finishes there is a signal sent to the Scanner to start the scan and recording. 
* the function to interate trough the stack of phase patterns is called - `iterate_scan_stack`
The problem is tha the function which iterates trough the images basically runs after the NIDAQ starts nad I don't know how to ensure that 
* first the scan stck reaches calling the command to choose the first phase pattern
* second the NIDAQ starts the scan
Also I would like my code to be be logical and do not destroy what there already is now.
![Image (2)](https://github.com/ImSwitch/ImSwitch/assets/39549241/895a83f5-8c36-489c-8582-67f3dffd202a)

**Reply by xcasas** [reply]
Ok, sorry I wrote something but I changed my mind.

I understand you did something similar but I was thinking if you trigger the process using the SLM Controller you would fix the issue.
But then I remembered, you can have your own ScanController implementation.
Why don't you have one that when it runs the scan it waits from the SLM signal? (not actively waiting but you can do so that when the button is clicked you do your preparation, and then it asynchronously starts with a signal from the communicationChannel triggered by the SLM controller?)

For example @jonatanalvelid did one for the point scanning: https://github.com/ImSwitch/ImSwitch/blob/master/imswitch/imcontrol/controller/controllers/ScanControllerPointScan.py
how would you do this @jonatanalvelid ?

**Reply by Kubiczek36** [reply]
Hi, sorry, that I haven't replied earlier. aAt the end we've fixed the issue, but the problem with the lack of paralelisation remains. If I Just run a scan like thisfrom the Scanning widget (it calls the same function as the rec. manager) I cannot see the camera updating, which is not good as I would like to see, what is happening with the beam. This is problem especially while I am trigering the camera (I don't see the trigered frame on the image display).  How can I ensure, that the scanning part, the SLM part and the Image display part run in paralell and I can see, what is happening.
